### DISCUSSION TITLE
Spawning new chromium windows while watching a full screen video in chromium makes the newly spawned window to not be maximized

### DISCUSSION THREAD

**Discussion by charbelnicolas** [labwc/labwc#2036] [Q&A]
When I'm watching a fullscreen video in chromium, if I spawn a new chromium instance, the newly spawned window will be unmaximized but centered on the screen. On top of that, if I now maximize said window, it won't center it on the screen but will be offset to the top left corner by a 1/3 of an inch and some window content will be cut off.

When launched:

![ss_2024_07_29_10_16_04](https://github.com/user-attachments/assets/0adc5ebb-ad95-40ee-b1e6-f1424a6883bb)

When maximized:

![ss_2024_07_29_10_16_08](https://github.com/user-attachments/assets/a30faae4-e12b-49a8-bef9-a666a1ef818f)

This only happens when launching chromium windows, but it won't happen with inkscape, blender or a terminal (from what I have tested).

Any ideas?

**Comment by charbelnicolas** [comment]
The black and white stripes in the background is a full screen video from youtube by the way.

**Comment by tokyo4j** [comment]
Thanks for reporting.

The first problem (newly spawned window is not maximized while fullscreen window is opened) doesn't look like our bug as I experienced the same behavior in KWin. When I run chromium with `WAYLAND_DEBUG=client`, it firstly sends `unset_maximized` request.

However, I don't know why the second problem (chromium window is incorrectly positioned when maximized) occurs. But as it looks like the offset is the same as the offset of CSD shadow (16x10), chances are Chromium fails to send the correct window geometry with `set_window_geometry` request or labwc is missing it for some reason.

As always, output of `WAYLAND_DEBUG=client chromium` will be helpful. Since windows are opened from a single process in Chromium, you will need to open the first window (in which you watch fullscreen video) with `WAYLAND_DEBUG=client chromium`.

**Reply by charbelnicolas** [reply]
Sure, here's the log, it's awfully long, like 11k lines:

[chromium.txt](https://github.com/user-attachments/files/16421448/chromium.txt)

**Reply by tokyo4j** [reply]
Could you give me the output when you maximize the newly spawned window? I want to investigate the second problem.

EDIT: Or did you actually maximized the newly spawned window? Then it's strange that I couldn't find `set_maximized` request.

EDIT2: If you maximized the newly spawned window, how did you do it? By clicking the maximize button on Chromium, by a keybind, or by a window rule?

**Reply by charbelnicolas** [reply]
I maximize it with this keybind:

```
<keybind key="W-S-m">
	<action name="Maximize" />
</keybind>
```
I did maximize the newly spawned window.

**Reply by charbelnicolas** [reply]
I can only find one instance of `[ 556611.506]  -> xdg_toplevel@37.set_maximized()` on line 464 but that might be the first window launch. I can't find any other. Maybe that is why it is not maximizing properly?

**Reply by charbelnicolas** [reply]
I think the problem is that when I send the maximize command chromium tries to full screen the window instead of maximizing it:

`[ 564925.657]  -> xdg_toplevel@37.set_fullscreen(nil)` on line 9648.

It is easier to replicate the issue simply by running `WAYLAND_DEBUG=client chromium --start-fullscreen` and then spawning a new window.

**Reply by charbelnicolas** [reply]
Here's the log from running `WAYLAND_DEBUG=client chromium --start-fullscreen`

[chromium_fullscreen.txt](https://github.com/user-attachments/files/16422094/chromium_fullscreen.txt)

Please let me know if you can see the video or if I should upload it to google drive.

https://github.com/user-attachments/assets/7a9ea8a5-2af2-4638-9a45-d069fe44782b

**Reply by charbelnicolas** [reply]
So I think this might be a sole chromium issue and not labwc at all or who knows -_-

**Reply by tokyo4j** [reply]
> Then it's strange that I couldn't find `set_maximized` request.

Sorry this was wrong. `set_maximized` request is sent by client when it wants to be maximized, but not when it's maximized by labwc's actions.

> I think the problem is that when I send the maximize command chromium tries to full screen the window instead of maximizing it:

I guess this is just a Chromium's issue and I reproduced it. But what I cannot reproduce is the incorrect window positioning that shifts the window to top left.

Below is the log when I run `WAYLAND_DEBUG=client --start-fullscreen chromium 2>&1 | grep -e xdg -e commit -e attach`.
In my setup, labwc sends `configure` to notify Chromium of the new window gemetry, and Chromium responds with `set_window_geometry(0, 0, 1920, 1080)` requests. But I couldn't find it in your log and I guess that's causing the incorrect window positioning.

```
[3880112.315] xdg_toplevel@51.configure(1920, 1080, array[4])
[3880112.323] zxdg_toplevel_decoration_v1@63.configure(1)
[3880112.325] xdg_surface@53.configure(6422)
[3880113.543]  -> xdg_toplevel@51.set_min_size(500, 25)
[3880113.549]  -> xdg_toplevel@51.set_max_size(0, 0)
[3880114.093]  -> xdg_toplevel@51.set_min_size(500, 35)
[3880114.097]  -> xdg_toplevel@51.set_max_size(0, 0)
[3880131.300]  -> wl_surface@52.commit()
[3880150.726]  -> xdg_surface@53.set_window_geometry(0, 0, 1920, 1080)
```

I have no clue why Chromium misbehaves like that in your setup. I'll look into your log more, but I'm not sure I can find the cause.

**Comment by charbelnicolas** [comment]
Here is the log from `WAYLAND_DEBUG=client chromium --start-fullscreen 2>&1 | grep -e xdg -e commit -e attach`

[chromium.txt](https://github.com/user-attachments/files/16444954/chromium.txt)

**Reply by charbelnicolas** [reply]
`[4108810.093]  -> xdg_surface@43.set_window_geometry(22, 19, 1876, 1036)`

I have a 1920 x 1080 display, so it is setting it wrong I guess...

**Reply by charbelnicolas** [reply]
Ok, I don't know what's up but it is magically working now... This was so random... I did update chromium to the latest 127 (I already had version 127, I just updated it to the very latest) version today on arch but I don't think that had to do anything with it...

[chromium_fixed.txt](https://github.com/user-attachments/files/16445025/chromium_fixed.txt)

**Comment by charbelnicolas** [comment]
I also ran this keybind:

```
<keybind key="W-S-f">
	<action name="ResizeTo" width="1920" height="1080" />
	<action name="FocusOutput" output="HDMI-A-1" />
	<action name="AutoPlace" policy="center" />
</keybind>
```
So that must have been it...

**Reply by charbelnicolas** [reply]
Sorry if this was a waste of your time, I don't know why it would offset it in the first place, maybe it has to do with cascading new windows?

**Comment by charbelnicolas** [comment]
What is funny now is that I get this message every time I launch chromium now:

`02:06:20.483 [ERROR] [../labwc/src/view.c:714] view has empty geometry, not centering`

**Comment by droc12345** [comment]
That checks for width and height being greater than zero.
To be more informative, the log should report the x,y and w,h parms along with the message

**Reply by charbelnicolas** [reply]
![ss_2024_08_01_11_46_11](https://github.com/user-attachments/assets/7f16dc9e-a2bc-41ce-83c8-7d69741adafe)

That is all I get, there is no x,y and w,h.

I also have this rule on my rc.xml:

`<windowRule identifier="*" serverDecoration="no" fixedPosition="yes" />`

I don't ever resize, drag or move windows, all I ever do is either maximize or fullscreen applications with keybinds.

**Reply by droc12345** [reply]
I mean the code that generates the log needs to add the x,y,w,h parms. (sorry about that)
